lib/randomfileaccess.js

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright 2020 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+"use strict";
+
+const PRIVATE = Symbol("PRIVATE");
+const { open } = require("fs").promises;
+
+/**
+ * @typedef {Object} WriteResponse
+ * @property {Number} bytesWritten the number of bytes written
+ * @property {Buffer} buffer the buffer written
+ */
+
+/**
+ * @typedef {Object} SubRange
+ * @property {Number} low Low-end of the range
+ * @property {Number} high High-end of the range (included)
+ * @property {Number} length Length of the range
+ */
+
+/**
+ * Open a file handle in the given mode, grow the file up to the expected target size
+ * 
+ * @param {String|URL} path File path
+ * @param {"r"|"w"} flags Mode to open the file
+ * @param {Number} [size] expected size of the file
+ * @returns {FileHandle} File handle
+ */
+async function openFile(path, flags, size) {
+    const handle = await open(path, flags);
+    try {
+        if ((flags === "w") && size) {
+            await handle.truncate(size);
+        }    
+        return { handle, flags, size };
+    } catch (err) {
+        await handle.close();
+        throw err;
+    }
+}
+
+/**
+ * Closes a file silently
+ * 
+ * @param {String|URL} path File path
+ * @param {Promise} fileHandlePromise File handle promise
+ */
+async function closeFile(path, fileHandlePromise) {
+    try {
+        const { handle } = await fileHandlePromise;
+        await handle.close();    
+    } catch (err) {
+        console.error(`Unable to close file handle of ${path}`, err);
+    }
+}
+
+/**
+ * Random file access 
+ * 
+ * This class holds the file handles for both "read" (upload) and "write" (download) so 
+ * independent {Transfer} function instances can read and write the same file concurrently.
+ */
+class RandomFileAccess {
+    /**
+     * Construct Random File Access
+     */
+    constructor() {
+        this[PRIVATE] = {
+            fileHandles: new Map()
```

Comment:
Note that file handles are *usually* limited by the OS. On Linux, `ulimit` can show that limit. I think it would be nice to add some handling in case there are a lot of file handles open, especially since I think this can happen for larger uploads/transfers (be it large files or a lot of small files).